From aaf77cb98e4c7ba4d775923444e5f0d93ae0e5d3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tim=20R=C3=BChsen?= <tim.ruehsen@gmx.de>
Date: Mon, 10 Aug 2026 17:21:30 +0200
Subject: [PATCH] Fix segfault in retrieve_from_metalink() when a metalink:file
 has no resources

* src/main.c (main): Guard access to mfile->resources.
* src/metalink.c (retrieve_from_metalink): Likewise.

Reported-by: vlefebvre <valentin.lefebvre@suse.com>
Co-authored-by: vlefebvre <valentin.lefebvre@suse.com>
---
 src/main.c     | 20 ++++++++++++--------
 src/metalink.c | 28 +++++++++++++++++++---------
 2 files changed, 31 insertions(+), 17 deletions(-)

diff --git a/src/main.c b/src/main.c
index 3926f501..bf4aed04 100644
--- a/src/main.c
+++ b/src/main.c
@@ -2224,17 +2224,21 @@ only if outputting to a regular file.\n"));
               metalink_file_t **mfile_ptr;
               for (mfile_ptr = metalink->files; *mfile_ptr; mfile_ptr++)
                 {
-                  metalink_resource_t **mres_ptr;
                   metalink_file_t *mfile = *mfile_ptr;
-                  size_t mres_count = 0;
 
-                  for (mres_ptr = mfile->resources; *mres_ptr; mres_ptr++)
-                    mres_count++;
+                  if (mfile->resources)
+                    {
+                      metalink_resource_t **mres_ptr;
+                      size_t mres_count = 0;
 
-                  stable_sort (mfile->resources,
-                               mres_count,
-                               sizeof (metalink_resource_t *),
-                               metalink_res_cmp);
+                      for (mres_ptr = mfile->resources; *mres_ptr; mres_ptr++)
+                        mres_count++;
+
+                      stable_sort (mfile->resources,
+                                  mres_count,
+                                  sizeof (metalink_resource_t *),
+                                  metalink_res_cmp);
+                    }
                 }
             }
           retr_err = retrieve_from_metalink (metalink);
diff --git a/src/metalink.c b/src/metalink.c
index 954546d2..4c228eb2 100644
--- a/src/metalink.c
+++ b/src/metalink.c
@@ -300,17 +300,21 @@ retrieve_from_metalink (const metalink_t* metalink)
                     metalink_file_t **x_mfile_ptr;
                     for (x_mfile_ptr = metaurl_xml->files; *x_mfile_ptr; x_mfile_ptr++)
                       {
-                        metalink_resource_t **x_mres_ptr;
                         metalink_file_t *x_mfile = *x_mfile_ptr;
-                        size_t mres_count = 0;
 
-                        for (x_mres_ptr = x_mfile->resources; *x_mres_ptr; x_mres_ptr++)
-                          mres_count++;
+                        if (x_mfile->resources)
+                          {
+                            metalink_resource_t **x_mres_ptr;
+                            size_t mres_count = 0;
 
-                        stable_sort (x_mfile->resources,
-                                     mres_count,
-                                     sizeof (metalink_resource_t *),
-                                     metalink_res_cmp);
+                            for (x_mres_ptr = x_mfile->resources; *x_mres_ptr; x_mres_ptr++)
+                              mres_count++;
+
+                            stable_sort (x_mfile->resources,
+                                         mres_count,
+                                         sizeof (metalink_resource_t *),
+                                         metalink_res_cmp);
+                          }
                       }
                   }
 
@@ -370,7 +374,7 @@ retrieve_from_metalink (const metalink_t* metalink)
 
       /* Resources are sorted by priority.  */
       for (mres_ptr = mfile->resources;
-           *mres_ptr && mfile->checksums && !skip_mfile; mres_ptr++)
+           mres_ptr && *mres_ptr && mfile->checksums && !skip_mfile; mres_ptr++)
         {
           metalink_resource_t *mres = *mres_ptr;
           metalink_checksum_t **mchksum_ptr, *mchksum;
@@ -859,6 +863,12 @@ gpg_skip_verification:
             } /* endif RETR_OK.  */
         } /* Iterate over resources.  */
 
+      if (!mfile->resources)
+        {
+          logprintf (LOG_NOTQUIET, _("No resources found.\n"));
+          retr_err = METALINK_MISSING_RESOURCE;
+        }
+
       if (!mfile->checksums)
         {
           logprintf (LOG_NOTQUIET, _("No checksums found.\n"));
-- 
2.55.0

